### PR TITLE
Implement geofence-based presence

### PR DESCRIPTION
**Pull request by jjhuff**
This implements #343 

* Added prefs.
* If Lat/Lng aren't set, it fetches `/services/org.eclipse.smarthome.core.i18nprovider/config` to get `location`. I'd be totally in favor of fetching that another way if there is one. Maybe a new core service to expose for location?
* Request permissions dynamically if they weren't already granted
* Setup geofences (on boot too!)
* Update item state on enter/exit

I don't do tons of Java or Android, so please tell me what can be improved!

### CODE REVIEW

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+
+        // Get a PendingIntent containing the entire back stack.
+        PendingIntent notificationPendingIntent =
+                stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT);
+
+        // Get a notification builder that's compatible with platform versions >= 4
+        NotificationCompat.Builder builder = new NotificationCompat.Builder(this);
+
+        // Define the notification settings.
+        builder.setSmallIcon(R.drawable.ic_launcher)
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java]: Usually apps use their icon with a background in notifications

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+                // In a real app, you may want to use a library like Volley
+                // to decode the Bitmap.
+                .setLargeIcon(BitmapFactory.decodeResource(getResources(),
+                        R.drawable.ic_launcher))
+                .setColor(Color.RED)
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java]: Why red?

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+                .setContentTitle(notificationDetails)
+                .setContentText("Geofence transition")
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java]: Please use strings.xml ressources for this.

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+    private String getTransitionString(int transitionType) {
+        switch (transitionType) {
+            case Geofence.GEOFENCE_TRANSITION_ENTER:
+                return "GEOFENCE_TRANSITION_ENTER";
+            case Geofence.GEOFENCE_TRANSITION_EXIT:
+                return "GEOFENCE_TRANSITION_EXIT";
+            case Geofence.GEOFENCE_TRANSITION_DWELL:
+                return "GEOFENCE_TRANSITION_DWELL";
+            default:
+                return "Unknown type"+transitionType;
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java]: Is this visible to the user? If so please use strings.xml ressources for this.

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceRegistrationService.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    public void onConnected(Bundle connectionHint) {
+        Log.d(TAG, "onConnected");
+
+        float lat;
+        float lng;
+        try {
+            lat = Float.parseFloat(mSettings.getString(Constants.PREFERENCE_PRESENCE_LAT, "0"));
+            lng = Float.parseFloat(mSettings.getString(Constants.PREFERENCE_PRESENCE_LNG, "0"));
+        } catch(NumberFormatException e) {
+            Log.i(TAG, "Invalid lat/lng");
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/GeofenceRegistrationService.java]: Does the user get feedback that lat/lng is invalid?

File: mobile/src/main/java/org/openhab/habdroid/core/OpenHABIntentService.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+        Log.d(TAG, "onError(): " + error);
+        stopSelf();
+    }
+
+
+    protected void processIntent(Intent data) {
+        Log.d(TAG, "processIntent()");
+        if (mOpenHABBaseUrl == null) {
+            Log.w(TAG, "Couldn't determine OpenHAB URL");
+            showToast("Couldn't determine OpenHAB URL");
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/OpenHABIntentService.java]: Please use strings.xml ressources for this.

File: mobile/src/main/java/org/openhab/habdroid/core/OpenHABIntentService.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+                        command, "text/plain;charset=UTF-8", new MyHttpClient.ResponseHandler() {
+                            @Override
+                            public void onSuccess(Call call, int statusCode, Headers headers, byte[] responseBody) {
+                                Log.d(TAG, "Command was sent successfully");
+                            }
+
+                            @Override
+                            public void onFailure(Call call, int statusCode, Headers headers, byte[] responseBody, Throwable error) {
+                                if (statusCode == 404) {
+                                    showToast("Item " + itemName + " doesn't exist");
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/OpenHABIntentService.java]: Please use strings.xml ressources for this.

File: mobile/src/main/java/org/openhab/habdroid/core/OpenHABIntentService.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+                                } else {
+                                    showToast("Error " + statusCode + " while sending " + command + " to " + itemName);
```
Comment (mueller-ma) [inline mobile/src/main/java/org/openhab/habdroid/core/OpenHABIntentService.java]: Please use strings.xml ressources for this.

File: mobile/src/main/res/xml/preferences.xml [hunk 9]
```diff
@@ -58,6 +58,40 @@
             android:summary="@string/settings_openhab_sslhost_summary"
             android:title="@string/settings_openhab_sslhost" />
     </PreferenceCategory>
+    <PreferenceCategory android:title="Presence">
+        <CheckBoxPreference
+            android:defaultValue="false"
+            android:key="default_openhab_presence_enable"
+            android:summary="Enable location-based presence reporting"
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: Please use strings.xml ressources for this.

File: mobile/src/main/res/xml/preferences.xml [hunk 10]
```diff
@@ -58,6 +58,40 @@
...
+            android:title="Enable" />
+        <EditTextPreference
+            android:defaultValue="isPresent"
+            android:dependency="default_openhab_presence_enable"
+            android:inputType="textNoSuggestions"
+            android:key="default_openhab_presence_item"
+            android:summary=""
+            android:title="Presence Item" />
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: Please use strings.xml ressources for this.

File: mobile/src/main/res/xml/preferences.xml [hunk 11]
```diff
@@ -58,6 +58,40 @@
...
+        <EditTextPreference
+            android:defaultValue=""
+            android:dependency="default_openhab_presence_enable"
+            android:inputType="numberDecimal"
+            android:key="default_openhab_presence_lat"
+            android:summary=""
+            android:title="Latitude" />
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: Please use strings.xml ressources for this.

File: mobile/src/main/res/xml/preferences.xml [hunk 12]
```diff
@@ -58,6 +58,40 @@
...
+        <EditTextPreference
+            android:defaultValue=""
+            android:dependency="default_openhab_presence_enable"
+            android:inputType="numberDecimal"
+            android:key="default_openhab_presence_lng"
+            android:summary=""
+            android:title="Longitude" />
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: Please use strings.xml ressources for this.

File: mobile/src/main/res/xml/preferences.xml [hunk 13]
```diff
@@ -58,6 +58,40 @@
...
+        <CheckBoxPreference
+            android:defaultValue="false"
+            android:dependency="default_openhab_presence_enable"
+            android:key="default_openhab_presence_debug"
+            android:summary="Enable debug notifications"
+            android:title="Debugging" />
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: Please use strings.xml ressources for this.

File: mobile/src/main/res/xml/preferences.xml [hunk 14]
```diff
@@ -58,6 +58,40 @@
...
+            android:title="Presence Item" />
+        <EditTextPreference
+            android:defaultValue=""
+            android:dependency="default_openhab_presence_enable"
+            android:inputType="numberDecimal"
+            android:key="default_openhab_presence_lat"
+            android:summary=""
+            android:title="Latitude" />
+        <EditTextPreference
+            android:defaultValue=""
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: You could replace "" with "@string/empty_string"

**Comment by digitaldan** [comment]
Hi @jjhuff , thanks for the PR, and thanks @mueller-ma for the initial review.  I need to find a large block of time to get through all the android PR's this week.

File: mobile/src/main/res/xml/preferences.xml (hunk 14 above)
Comment (jjhuff) [reply mobile/src/main/res/xml/preferences.xml]: Not entire clear what you meant, but I just removed the default.

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceRegistrationService.java (hunk 5 above)
Comment (jjhuff) [reply mobile/src/main/java/org/openhab/habdroid/core/GeofenceRegistrationService.java]: They do now:)

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java (hunk 1 above)
Comment (jjhuff) [reply mobile/src/main/java/org/openhab/habdroid/core/GeofenceIntentService.java]: Indeed! This was lifted from elsewhere. Fixed.

**Comment by jjhuff** [comment]
Thanks for the review @mueller-ma. I think I fixed everything you flagged.

File: mobile/src/main/res/xml/preferences.xml [hunk 15]
```diff
@@ -58,6 +58,38 @@
             android:summary="@string/settings_openhab_sslhost_summary"
             android:title="@string/settings_openhab_sslhost" />
     </PreferenceCategory>
+    <PreferenceCategory android:title="Presence">
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: Hardcoded string

File: mobile/src/main/res/xml/preferences.xml [hunk 16]
```diff
@@ -58,6 +58,38 @@
...
+            android:dependency="default_openhab_presence_enable"
+            android:inputType="numberDecimal"
+            android:key="default_openhab_presence_lat"
+            android:summary=""
+            android:title="@string/presence_pref_lat_title" />
+        <EditTextPreference
+            android:dependency="default_openhab_presence_enable"
+            android:inputType="numberDecimal"
+            android:key="default_openhab_presence_lng"
+            android:summary=""
```
Comment (mueller-ma) [inline mobile/src/main/res/xml/preferences.xml]: Can you display the selected value as summary?

Comment (jjhuff) [reply mobile/src/main/res/xml/preferences.xml]: That's a good idea, done!

**Comment by digitaldan** [comment]
@jjhuff I just had a chance to run this. Couple of suggestions.  My thoughts for this functionality would be that you select a location from a map (using google map api) and then select and item to switch on when you enter it, off when you exit it.  I was not expecting the user to manually enter a single lat/lon or the actual item name.   Would it be possible to let the user select a map for the location, and also fetch a list of items (/rest/items) from the users server to select?  Even better, supporting multiple geofences -> items would be awesome.

**Comment by jjhuff** [comment]
@digitaldan I think both of those would be great, but I wanted to start somewhere

* Selecting items: This might be a little thorny, but I need to dig into the preferences code some more. I'm not sure we'd always have server info available (especially when initially configuring the app).

* As for the map: I think the best option is for the server to provide the location(s) in most (maybe all) cases. I implemented that by pulling `location` from `i18nprovider`, but I'd prefer a better option. Failing all of that, I agree 100% that selecting from a map is preferable to entering LatLng.

**Comment by digitaldan** [comment]
>Selecting items: This might be a little thorny, but I need to dig into the preferences code some more. I'm not sure we'd always have server info available (especially when initially configuring the app).

What I would like to see is this not in the preferences menu, but in the main menu where we already have "write nfc tag", so something like "Geo Fences".  when selecting this it pops up a window showing your saved geofences , or a button/link to create a new one.  

For selecting an item, If you are not connected to a server then you would have an empty dropdown menu, but could still type into the field (if we think its an issue) , selecting from the dropdown would populate that item name in the field. 

>As for the map: I think the best option is for the server to provide the location(s) in most (maybe all) cases. I implemented that by pulling location from i18nprovider, but I'd prefer a better option. Failing all of that, I agree 100% that selecting from a map is preferable to entering LatLng.

So, I think your point of view here is that this is meant for geofencing your home, but that is only one use case.  For example, I have a geofence when I drive into my neighborhood and another one for when I get home.  I also have a geo fence for work.   Tying this to a single location seems a bit limiting, and not what I would expect from a geo fence feature.

Btw, I love this feature!  We want to bump our app version to 2.0 soon and this could be a great reason to do so.

**Comment by jjhuff** [comment]
> What I would like to see is this not in the preferences menu, but in the main menu where we already have

Gotcha. Yup, that solves a lot of problems.

>  Tying this to a single location seems a bit limiting, and not what I would expect from a geo fence feature.

100% agree! However, my main point is that I think that the fences should be defined/stored on the server rather than every single client. Perhaps I have multiple devices, a family, or employees.

Proposal 1:
What if we made a service that does nothing but export a list of geofences similar to how the OwnTracks stuff is configured

Proposal 2:
Perhaps better, configure all of it in the item. Then the app just needs to select a list of geofence items to update!
```
Switch  PresenceBen_PhoneHome   "Ben @ Home"   { geofence="-47.456,122.366,100" }
Switch  PresenceBen_PhoneWork   "Ben @ Work"   { geofence="-47.134,122.655,100" }
```
I suppose you could also have a config file to assign symbolic names to replace the lat,lng,radius values.

**Comment by digitaldan** [comment]
Apologies for the very late reply, I am catching up on PR's today.  

So i had to think about this, I still would prefer that a user selects a location and radius from a map, then selects an item.  I think this would be amazingly simple and be used frequently.

I am however trying to also see it from your POV, and so keeping an open mind.  

 For this project, we always have to keep in mind that it's the official Android app for openHAB.  This means we need to keep design focused on complimenting the openHAB server and its built in functionality.  So I'm generally opposed to requiring users to have to install a new binding unless it would absolutely be necessary .  I also believe we can not overload basic OH concepts like items or sitemaps to meet bespoke functionality.  

That being said, we do have the concept of a Location item.  If the user were to select a location item and another item to switch on or off I have a couple questions on how this would work.

1) how would we select and store the geofence radius? 
2) what happens when the user changes this on the server?  

>100% agree! However, my main point is that I think that the fences should be defined/stored on the server rather than every single client. Perhaps I have multiple devices, a family, or employees.

My opinion is that this over complicates the feature, but again its' just my opinion.  

>Proposal 1:
What if we made a service that does nothing but export a list of geofences similar to how the OwnTracks stuff is configured

The android app should work without any special configuration on the server that is specific to the app, i would like to avoid the user having to install a binding unless there was no alternative.

>Proposal 2:
Perhaps better, configure all of it in the item. Then the app just needs to select a list of geofence items to update!

I think we are overloading what an item is, so I don't think we can do that in this app. 

I'm hoping you are still interested in developing this feature!  I think it would be awesome.

**Comment by lolodomo** [comment]
Just an idea: why not having something similar to the voice recognition, that is the Android app updating a location item with the current phone position ? Then the user could use this position in rules and ser switch for example when the phone position is around a particular position.
Of course, the item name has to be a setting in the Android app because we must consider and differenciate several phones.

With which frequence the app will update the position ? With what impact on phone battery ?

**Comment by digitaldan** [comment]
>With which frequence the app will update the position ? With what impact on phone battery ?

I think that's the issue, not only would our app now wake up in the background for location changes events, but it also tries to wake up the radio and make REST calls at some frequent interval.  I was hoping not to have to deal with possible battery issues with our app.

**Comment by jjhuff** [comment]
I apologize for being MIA -- new job sucked away all of my free time!

How are people feeling about where this sits? I see an approval (I missed that earlier). I'm happy to clean up the merge conflicts if there's still overall interest in the feature.

I agree that a map would be a logical next step, but I'm also not super keen to draw this out.

Thoughts?

**Comment by digitaldan** [comment]
>Having the data stored on client isnt very practical, especially since you dont have a map to pick the location.

Maybe I don't understand, apps that I use that do geo fencing pop up a google map for me to select a location and radius, so we can have client access to a map, yes? 

Also on Android you have to register for entry/exits events to geo-fence locations, which implies the client does need to store the coordinates and some item to trigger locally.  In fact, I believe on each reboot of the device, apps need to listen for a boot complete action (or something) and re-register its geofences.

**Comment by jjhuff** [comment]
FWIW, I'm actually a big fan of having the geofences come entirely from the server -- I'm just not sure of the best mechanism to make that happen.

**Comment by mueller-ma** [comment]
Maybe as state of a string item, but thats not the best solution

**Comment by mueller-ma** [comment]
I really like proposal 2 of @jjhuff, but that would need some work on the server, wouldn't it?

**Comment by mueller-ma** [comment]
Triggering travis

**Comment by kubawolanin** [comment]
Looking forward to test this!
Is there anything missing on this PR (apart frorm the conflicts) to have it merged to the master branch?
Best regards

**Comment by mueller-ma** [comment]
Currently you have to enter the location's coordinates. It would be a lot nicer to fetch them from the server.

File: mobile/src/main/java/org/openhab/habdroid/core/GeofenceBroadcastReceiver.java [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+    public void onReceive(Context context, Intent intent) {
+        Log.d(TAG, "onReceive");
+        if(intent.getAction()!= null && intent.getAction().equals(Intent.ACTION_BOOT_COMPLETED)){
+            // Did we just boot? Go register.
+            Intent serviceIntent = new Intent(context , GeofenceRegistrationService.class);
+            startWakefulService(context, serviceIntent);
+        } else {
+            ComponentName comp = new ComponentName(context.getPackageName(),
+                    GeofenceIntentService.class.getName());
+            startWakefulService(context, (intent.setComponent(comp)));
```
Comment (maniac103) [inline mobile/src/main/java/org/openhab/habdroid/core/GeofenceBroadcastReceiver.java]: The if and the else block are 100% equivalent.

**Comment by lolodomo** [comment]
I just propose to implement a new geofence service in the server explaning how any UI could interact with.
https://github.com/eclipse/smarthome/issues/5066
I am waiting for your comments.

**Comment by lolodomo** [comment]
My proposal has the advantage to have the geofences defined on server side. Using Paper UI, it will be easy to select the locations insde a map (coming feature in openHAB).
By the way, the Android app could then display all the geofences in a map too, to help the user select the one to use.
The other difference is that the computation whether device is inside geofence or not, is done by the server, the UI just has to call the server with its current position, the selected geofence and an item to update.
WDYT ?

**Comment by maniac103** [comment]
I see that (computation being done server side) as a problem though. When doing it client side, one can leverage OS features (here: Play Services) when determining geofence status. When doing it server side, there's no other option than regularly polling location, which likely is less efficient.

**Comment by lolodomo** [comment]
No problem, we can have a service mode where the geofence status is determined by the client. In this case, the client only call the geofence service when it detects a change of status; the service will not compute the status again but consider the status provided in parameter.

**Comment by lolodomo** [comment]
In case the app only call the server when it enters or leaves a geo-fence, we have no chance to track the position of the person inside openHAB server. Maybe the service could be called by the Android app every XX minutes if the position has changed in order to set a location item. XX should be configurable.

**Comment by maniac103** [comment]
A geofence is not the same thing as location tracking. The former only provides an in/out information, the latter provides the actual location. We could probably support both modes, but the latter rather sounds like an extension of Location items.

**Comment by lolodomo** [comment]
Location tracking could already be done by any client UI simply by updating a particular location item. That is true that no additional service is required for that.

I thought about location tracking because if we have a service that receives the position for geofence computation, it is easy to track position at the same time. But that would be just an option of the service.

**Comment by lolodomo** [comment]
At least the service could allow sharing geofences. It could be used to save on server side new geofences defined by a client.

**Comment by kaikreuzer** [comment]
> @kaikreuzer rejected my proposal

... sounds a bit hard, I would say I questioned the suggested way to implement it.

But I think implementing this in the client side only isn't too bad as the precision and size of the fence might anyhow depend on the OS.

**Comment by mueller-ma** [comment]
Superseded by #914

**Pull request closed without merging**